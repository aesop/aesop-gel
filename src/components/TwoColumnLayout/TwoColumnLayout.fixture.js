import React from 'react';

export default {
  childrenClassNames: { content: 'content', sidebar: 'sidebar' },
  className: 'TwoColumnLayout',
  content: <div>content</div>,
  hasRightPadding: true,
  hasVerticalPadding: false,
  id: 'TwoColumnLayout',
  isReversed: true,
  sidebar: <div>sidebar</div>,
};
