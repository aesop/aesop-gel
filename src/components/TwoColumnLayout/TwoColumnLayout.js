import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import styles from './TwoColumnLayout.module.css';

const TwoColumnLayout = ({
  childrenClassNames,
  className,
  content,
  hasFullWidthContent,
  id,
  isFlushOnSmall,
  isReversed,
  sidebar,
  theme,
}) => {
  const classSet = cx(
    styles.base,
    { [styles.reverse]: isReversed },
    { [styles.contentOnly]: !sidebar },
    styles[theme],
    className,
  );
  const contentClassSet = cx(
    styles.content,
    { [styles.fullWidth]: hasFullWidthContent },
    { [styles.flushOnSmall]: isFlushOnSmall },
    childrenClassNames.content,
  );
  const sidebarClassSet = cx(styles.sidebar, childrenClassNames.sidebar);

  return (
    <div className={classSet} id={id}>
      <div className={contentClassSet} id="mainColumn">
        {content}
      </div>
      {sidebar && (
        <aside className={sidebarClassSet} id="sidebarColum">
          {sidebar}
        </aside>
      )}
    </div>
  );
};

TwoColumnLayout.propTypes = {
  childrenClassNames: PropTypes.shape({
    sidebar: PropTypes.string,
    content: PropTypes.string,
  }),
  className: PropTypes.string,
  content: PropTypes.any.isRequired,
  hasFullWidthContent: PropTypes.bool,
  id: PropTypes.string,
  isFlushOnSmall: PropTypes.bool,
  isReversed: PropTypes.bool,
  sidebar: PropTypes.any,
  theme: PropTypes.oneOf(['dark', 'light']),
};

TwoColumnLayout.defaultProps = {
  childrenClassNames: { sidebar: undefined, content: undefined },
  className: undefined,
  content: undefined,
  hasFullWidthContent: false,
  id: undefined,
  isFlushOnSmall: false,
  isReversed: true,
  sidebar: undefined,
  theme: 'dark',
};

export default TwoColumnLayout;
