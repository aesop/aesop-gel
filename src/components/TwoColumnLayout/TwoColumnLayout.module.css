@import 'styles/index';

.base {
  display: flex;
  box-sizing: border-box;
  flex-wrap: wrap;
  justify-content: center;

  * {
    &,
    &::before,
    &::after {
      box-sizing: inherit;
    }
  }

  @media (--viewport-lg) {
    justify-content: flex-start;
  }
}

.sidebar {
  width: calc(100% - var(--layout-sm-spacing) * 2);
  margin-bottom: calc(var(--layout-sm-spacing) * 2);

  @media (--viewport-md) {
    width: calc((100% - var(--layout-md-spacing) * 2) * 5 / 6);
    margin-bottom: calc(var(--layout-md-spacing) * 2);
  }

  @media (--viewport-lg) {
    width: calc((100% - var(--layout-lg-spacing) * 2) * 5 / 12);
    padding: 0 var(--layout-lg-spacing) 0 0;
    margin-right: calc((100% + var(--layout-lg-spacing) * 2) * 1 / 12);
    margin-bottom: 0;
  }
}

.content {
  width: calc(100% - var(--layout-sm-spacing) * 2);
  margin-bottom: calc(var(--layout-sm-spacing) * 2);

  @media (--viewport-md) {
    width: calc((100% - var(--layout-md-spacing) * 2) * 5 / 6);
    margin-bottom: calc(var(--layout-md-spacing) * 2);
  }

  @media (--viewport-lg) {
    width: calc((100% - var(--layout-lg-spacing)) * 6 / 12);
    margin-bottom: 0;
    margin-left: var(--layout-lg-spacing);
  }

  &.fullWidth {
    @media (--viewport-md) {
      width: calc((100% - var(--layout-md-spacing) * 2) * 5 / 6);
    }

    @media (--viewport-lg) {
      width: calc((100% - var(--layout-lg-spacing)) * 7 / 12);
    }

    &.reverse {
      /* reverse */
    }
  }
}

.reverse {
  flex-direction: column-reverse;
  align-content: center;

  @media (--viewport-lg) {
    flex-direction: row-reverse;
    justify-content: flex-end;
  }

  .content {
    @media (--viewport-lg) {
      margin-left: 0;
    }
  }

  .sidebar {
    @media (--viewport-lg) {
      padding: 0 calc((100% - var(--layout-lg-spacing) * 2) / 12) 0 0;
      margin-right: 0;
    }
  }
}

.contentOnly {
  @media (--viewport-lg) {
    justify-content: flex-start;
  }

  &.reverse {
    justify-content: flex-end;

    .content {
      @media (--viewport-lg) {
        width: calc((100% - var(--layout-lg-spacing)) * 6 / 12);
        margin-left: calc((100% - var(--layout-lg-spacing) * 2) * 5 / 12);
      }

      &.fullWidth {
        @media (--viewport-md) {
          width: calc((100% - var(--layout-md-spacing) * 2) * 5 / 6);
        }

        @media (--viewport-lg) {
          width: calc((100% - var(--layout-lg-spacing)) * 7 / 12);
        }
      }
    }
  }
}
