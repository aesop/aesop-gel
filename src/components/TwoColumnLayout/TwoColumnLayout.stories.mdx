import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import { withKnobs } from '@storybook/addon-knobs';
import { text, boolean, select } from '@storybook/addon-knobs';
import '~.storybook/storybook.module.css';
import TwoColumnList from '~/components/TwoColumnList';
import TwoColumnListFixture from '~/components/TwoColumnList/TwoColumnList.fixture';
import BodyCopy from '~/components/BodyCopy';
import TwoColumnLayout from './TwoColumnLayout';

<Meta
  title="Components/TwoColumnLayout"
  component={TwoColumnLayout}
  decorators={[withKnobs]}
/>

# TwoColumnLayout

<Preview>
  <Story name="Base component">
    <TwoColumnLayout
      childrenClassNames={{ sidebar: 'sidebar', content: 'content' }}
      content={
        <>
          <BodyCopy
            copy="Parsley Seed Cream. [The Paris Review](http://theparisreview.org) Intensely soothing, nourishing and hydrating, this elegant formulation contains a potent blend of botanicals that offer fortification of the highest order against free radicals."
            heading="Heading hasRightPadding isReversed Content Column nourishing and hydrating"
            id="Anti-Oxidant"
            theme={select('theme', ['dark', 'light'], 'dark')}
          />
          <TwoColumnList
            leftColumn={TwoColumnListFixture.leftColumn}
            rightColumn={TwoColumnListFixture.rightColumn}
            theme={select('theme', ['dark', 'light'], 'dark')}
          />
        </>
      }
      hasRightPadding={boolean('hasRightPadding', true)}
      hasVerticalPadding={boolean('hasVerticalPadding', true)}
      id={text('id', 'TwoColumn-id')}
      isReversed={boolean('isReversed', true)}
      sidebar={
        <BodyCopy
          copy="Parsley Seed Cream. [The Paris Review](http://theparisreview.org) Intensely soothing, nourishing and hydrating, this elegant formulation contains a potent blend of botanicals that offer fortification of the highest order against free radicals."
          cta={{
            id: 'button-id',
            text: 'Read more',
            url: 'http://aesop.com',
            openInANewWindow: true,
            style: 'Internal Button Link',
          }}
          eyebrow="Recommended Nearby"
          hasSerifFontHeading={true}
          heading="Heading hasRightPadding isReversed Sidebar Column"
          id="Anti-Oxidant"
          subHeading="Active Nutrients"
          theme={select('theme', ['dark', 'light'], 'dark')}
        />
      }
      theme={select('theme', ['dark', 'light'], 'dark')}
    />
  </Story>
</Preview>

## API

<Props of={TwoColumnLayout} />

## Usage Example

```
<TwoColumnLayout
  childrenClassNames={{ sidebar: '...', content: '...' }}
  content={</>}
  hasRightPadding={true}
  hasVerticalPadding={true}
  id="TwoColumn-id"
  isReversed={true}
  sidebar={</>}
/>
```
