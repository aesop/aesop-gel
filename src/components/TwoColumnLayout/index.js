/* istanbul ignore file */
import { TwoColumnLayout } from './TwoColumnLayout';

export { TwoColumnLayout };
