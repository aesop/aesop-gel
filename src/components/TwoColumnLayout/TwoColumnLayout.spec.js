import React from 'react';
import { render } from '@testing-library/react';
import TwoColumnLayout from './TwoColumnLayout';
import TwoColumnLayoutFixture from './TwoColumnLayout.fixture';

describe('<TwoColumnLayout />', () => {
  it('should be defined', () => {
    expect(TwoColumnLayout).toBeDefined();
  });

  it('renders base component correctly', () => {
    const { container } = render(
      <TwoColumnLayout
        childrenClassNames={TwoColumnLayoutFixture.childrenClassNames}
        className={TwoColumnLayoutFixture.className}
        content={TwoColumnLayoutFixture.content}
        hasFullWidthContent={TwoColumnLayoutFixture.hasFullWidthContent}
        id={TwoColumnLayoutFixture.id}
        isReversed={TwoColumnLayoutFixture.isReversed}
        sidebar={TwoColumnLayoutFixture.sidebar}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
