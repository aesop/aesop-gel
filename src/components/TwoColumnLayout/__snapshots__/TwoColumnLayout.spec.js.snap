// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TwoColumnLayout /> renders base component correctly 1`] = `
<div>
  <div
    class="base reverse dark TwoColumnLayout"
    id="TwoColumnLayout"
  >
    <div
      class="content fullWidth content"
      id="mainColumn"
    >
      <div>
        content
      </div>
    </div>
    <aside
      class="sidebar sidebar"
      id="sidebarColum"
    >
      <div>
        sidebar
      </div>
    </aside>
  </div>
</div>
`;
